flink-core/src/main/java/org/apache/flink/configuration/description/LineBreakElement.java

```diff
@@ -0,0 +1,40 @@
+/*
+ * Licensed to the Apache Software Foundation (ASF) under one
+ * or more contributor license agreements.  See the NOTICE file
+ * distributed with this work for additional information
+ * regarding copyright ownership.  The ASF licenses this file
+ * to you under the Apache License, Version 2.0 (the
+ * "License"); you may not use this file except in compliance
+ * with the License.  You may obtain a copy of the License at
+ *
+ *     http://www.apache.org/licenses/LICENSE-2.0
+ *
+ * Unless required by applicable law or agreed to in writing, software
+ * distributed under the License is distributed on an "AS IS" BASIS,
+ * WITHOUT WARRANTIES OR CONDITIONS OF ANY KIND, either express or implied.
+ * See the License for the specific language governing permissions and
+ * limitations under the License.
+ */
+
+package org.apache.flink.configuration.description;
+
+/**
+ * Represents a line break in the {@link Description}.
+ */
+public class LineBreakElement implements BlockElement {
+
+	/**
+	 * Creates a line break in the description.
+	 */
+	public static LineBreakElement linebreak() {
+		return new LineBreakElement();
+	}
+
+	private LineBreakElement() {
+	}
+
+	@Override
+	public String format(Formatter formatter) {
+		return formatter.format(this);
```

Comment:
hmm..................................................................

Alright so this works but I still think it's kinda weird that the element calls into the formatter; this effectively gives the element control over how it is formatted, since it _could_ just return an arbitrary string.

Let me sketch out an alternative, the general idea being to generalize the parent interface (DescriptionElement) to accommodate all sub-classes, and introduce an enum for categorization.
With this the elements are just a container for data, and most (but not all logic unfortunately unless we go for instanceof checks) formatting logic is in the formatter.
```
public enum ElementType {
	TEXT,
	LINK,
	LIST,
	LINE_BREAK,
	SEQUENCE // replaces your nested Text constructor
}

public interface Element {
	String getValue()
	List<Element> getChildren()
	ElementType getType()
}

class HtmlFormatter implements Formatter {

	format(Description description) {
		description.getElements().stream()
			.forEach(this::format)
			.collect(Collectors.joining())
	}

	String format(Element element) {
		switch (element.getType()) {
			case TEXT:
				return element.getValue()
			case LIST:
				StringBuilder sb = new StringBuilder()
				for (Element item : element.getChildren()) {
					sb.append(<whatver html/formatting prefix you use for lists>)
					sb.append(format(item))
					sb.append(<whatver html/formatting suffix you use for lists>)
				}
				return sb.toString()
			case LINK:
				return "<a href=" + element.getValue + ">"
			case LINE_BREAK:
				reutrn "<br>"
			case SEQUENCE:
				StringBuilder sb = new StringBuilder()
				for (Element item : element.getChildren()) {
					sb.append(format(item))
				}
				return sb.toString()
		}
	}
}

class Link implements Element {
	private final String value

	Link(String text) {
		this.value = text
	}

	getValue() {
			return this.value
	}

	getChildren() {
			return Collections.emptyList()
	}

	getType() {
		return LINK
	}
}

class Text implements Element {
	Text(String text) {
		this.value = text
		this.children = Collections.emptyList()
	}

	getValue() {
		return this.value
	}

	getChildren() {
		return Collections.emptyList
	}

	getType() {
		return TEXT
	}
}


class List implements Element {
	private final List<Element> elements

	List(Element ... elements) {
		this.elements = Arrays.asList(elements)
	}

	getValue() {
		return null
	}

	getChildren() {
		return elements
	}

	getType() {
		return LIST
	}
}

class LineBreak implements Element {

	getValue() {
		return null
	}

	getChildren() {
		return Collections.emptyList
	}

	getType() {
		return LINE_BREAK
	}
}

class ComboArrayWhatever implements Element {
	private final List<Element> elements

	List(Element ... elements) {
		this.elements = Arrays.asList(elements)
	}

	getValue() {
		return null
	}

	getChildren() {
		return Collections.emptyList
	}

	getType() {
		return SEQUENCE
	}
}

```

Reply:
this was only a rough draft, you could have inlineElement always return empty lists for getChildren and blockelement override getValue() and mark those as final. Finally `getChildren` could be changed to only return `InlineElement`, voila no infinite nesting. (see below for an updated draft)

Note that your approach currently has infinite nesting capabilities due to `Text`, which accepts a number of `InlineElements`, which again can be `Text`. (hence why i added the separate Sequence type)

The whole idea here is to _not_ allow elements to do super custom stuff. If we want to shorten links for the documentation we could have the formatter always show them as `here`.

```
public enum ElementType {
	TEXT,
	LINK,
	LIST,
	LINE_BREAK,
	ARRAY,
	SEQUENCE
}

public interface Element {
	String getValue()
	List<InlineElement> getChildren()
	ElementType getType()
}

class InlineElement implements Element {
	@Override
	final List<InlineElement> getChildren() {
		return Collections.emptyList;
	}
}

class BlockElement implements Element {
	@Override
	final String getValue() {
		return null;
	}
}

class HtmlFormatter implements Formatter {

	format(Description description) {
		description.getElements().stream()
			.forEach(this::format)
			.collect(Collectors.joining())
	}

	String format(Element element) {
		switch (element.getType()) {
			case TEXT:
				return element.getValue()
			case LIST:
				StringBuilder sb = new StringBuilder()
				for (Element item : element.getChildren()) {
					sb.append(<whatver html/formatting prefix you use for lists>)
					sb.append(format(item))
					sb.append(<whatver html/formatting suffix you use for lists>)
				}
				return sb.toString()
			case LINK:
				return "<a href=" + element.getValue + ">"
			case LINE_BREAK:
				reutrn "<br>"
			case SEQUENCE:
				StringBuilder sb = new StringBuilder()
				for (Element item : element.getChildren()) {
					sb.append(format(item))
				}
				return sb.toString()
		}
	}
}

class Link extends InlineElement {
	private final String value

	Link(String text) {
		this.value = text
	}

	getValue() {
			return this.value
	}

	getType() {
		return LINK
	}
}

class Text extends InlineElement {
	private final String value

	Text(String text) {
		this.value = text
	}

	getValue() {
			retu this.value;
		}

	getType() {
		return TEXT
	}
}


class List extends BlockElement {
	private final List<InlineElement> elements

	List(InlineElement ... elements) {
		this.elements = Arrays.asList(elements)
	}

	getChildren() {
		return elements
	}

	getType() {
		return LIST
	}
}

class LineBreak extends InlineElement {


	getValue() {
		return null
	}

	getType() {
		return LINE_BREAK
	}
}

class ComboArrayWhatever extends BlockElement {
	private final List<InlineElement> elements

	List(InlineElement ... elements) {
		this.elements = Arrays.asList(elements)
	}

	getChildren() {
		return Collections.emptyList
	}

	getType() {
		return SEQUENCE
	}
}
```

Reply:
As for the infinite nesting I could do exactly the same with the Visitor Pattern approach and introduce `Sequence` class. Both approaches are exactly the same in case of class hierarchy. It is just the question of how do we do a pattern matching - formatting. In languages with pattern matching on class type it is straightforward and we would go for a switch that would do a safe type inference.

In java we have two options Visitor pattern(my suggestion), switch on enum(your suggestion). I feel we are down to personal preference on that matter.

Why do I prefer the first one is that the classes are made to measure. They reflect structure of element: no empty methods, if unnecessary - e.g. no `getValue` in `LineBreak`, no `getChildren` methods in `InlineElements`, can have specialized fields in elements e.g. description for link, maybe importance for heading (h1-h6), could add number of line breaks into a single element.

I am afraid we end up in a preference lockdown.

Reply:
yes, nesting is an orthogonal issue.

For the nesting I suggest to change `Text` to implement `BlockElement` instead; it then effectively behaves like `Sequence`. At the moment i don't see a use-case for inserting Text into the format String since it could just be part of the format string.

I mean, when would this make sense?
```
text("hello %s", text("world"))
``

Reply:
Agree, it does not make sense. The reason why `Text` implements both `BlockElement` and `InlineElement` was to use `InlineElement` as entry in `ListElement`. This change allowed `ListElement` to contain both list and text.